In [7]:
import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
import tweepy
import io
import requests
import json
# Access Twitter API
consumer_key = 'OAuTFUGAfZ7jifVfGWg03S9cS'
consumer_secret = 'AEHFhdbvzIfXY6D1LWovUr0n95ovDZT7AICVd69UlpRJRi8LgQ'
access_token = '1320543920799969280-iXVCG7UvDtaaH95P2wSdk9CXnMW23e'
access_secret = 'cPOyzbf1eCHvjEtBb2qQmwaRtZiS80PQCTrxacyc18XwN'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [16]:
# WeRateDogs tweet archive
tweet_archive = pd.read_csv('/Users/jordanmazza/Desktop/Udacity Files/Project 4/WeRateDogs Project/twitter-archive-enhanced.csv')

In [15]:
# Tweet image predictions
url = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv').content
tweet_images = pd.read_csv(io.StringIO(url.decode('utf-8')), delimiter='\t')

In [10]:
# Retrieve JSON files from Tweet ID
tweet_list = []
exceptions = []
for tweet_id in tweet_archive.tweet_id:
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended')
        tweet_list.append({'tweet_id': int(tweet_id),
               'favorite_count': int(tweet.favorite_count), 
               'retweet_count': int(tweet.retweet_count),
    except Exception as e:
        exceptions.append(tweet_id)
        

print(exceptions)
print(tweet_list)

Rate limit reached. Sleeping for: 729
Rate limit reached. Sleeping for: 728


[888202515573088257, 873697596434513921, 872668790621863937, 872261713294495745, 869988702071779329, 866816280283807744, 861769973181624320, 856602993587888130, 851953902622658560, 845459076796616705, 844704788403113984, 842892208864923648, 837366284874571778, 837012587749474308, 829374341691346946, 827228250799742977, 812747805718642688, 802247111496568832, 779123168116150273, 775096608509886464, 771004394259247104, 770743923962707968, 759566828574212096, 754011816964026368, 680055455951884288]
[{'tweet_id': 892420643555336193, 'favorite_count': 35484, 'retweet_count': 7494, 'time': Timestamp('2017-08-01 16:23:56')}, {'tweet_id': 892177421306343426, 'favorite_count': 30716, 'retweet_count': 5562, 'time': Timestamp('2017-08-01 00:17:27')}, {'tweet_id': 891815181378084864, 'favorite_count': 23088, 'retweet_count': 3684, 'time': Timestamp('2017-07-31 00:18:03')}, {'tweet_id': 891689557279858688, 'favorite_count': 38760, 'retweet_count': 7667, 'time': Timestamp('2017-07-30 15:58:51')}, {'

In [75]:
#Dump JSON files to a .txt file
with open('tweet_json.txt', 'w') as file:
    for tweet in tweet_list:
        json.dump(tweet, file, default=str)
        file.write('\n')
    
tweets_array = []
with open ('tweet_json.txt', 'r') as txt_file:
    for line in txt_file:
        line_json = json.loads(line)
        tweets_array.append(line_json)

df_tweets = pd.DataFrame(tweets_array, columns = ['tweet_id','retweet_count', 'favorite_count'])
df_tweets.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7494           35484
1  892177421306343426           5562           30716
2  891815181378084864           3684           23088
3  891689557279858688           7667           38760
4  891327558926688256           8279           37014

## Assessing
### Visual Assessment

In [84]:
# Twitter Archive
tweet_archive.head()
tweet_archive.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [85]:
# Twitter Images
tweet_images.head()
tweet_images.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [86]:
# Tweet Dataframe
df_tweets.head()
df_tweets.tail()

tweet_id  retweet_count  favorite_count
2326  666049248165822465             40              96
2327  666044226329800704            125             266
2328  666033412701032449             39             110
2329  666029285002620928             41             120
2330  666020888022790149            454            2358

### Programmatic Assessment

In [89]:
# Twitter Archive
tweet_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [87]:
# Twitter Archive denominators
tweet_archive.rating_denominator.sort_values()
tweet_archive.query('rating_denominator != 10').shape

(23, 17)

In [88]:
# Twitter Archive numerators
tweet_archive.rating_numerator.sort_values()
tweet_archive.query('rating_numerator <=10 or rating_numerator >= 20').shape

(926, 17)

In [90]:
tweet_archive[tweet_archive['name'].duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
23    2017-07-19 00:47:34 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
23    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
23    This is Canela. She attempted some fancy porch...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
7                          NaN                        NaN   
12                         NaN                        NaN   
23                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
23    https://twitter.com/dog_rates/status/887473957...                13   
24    https://twitt

In [98]:
print(tweet_archive['doggo'].value_counts())
print(tweet_archive['floofer'].value_counts())
print(tweet_archive['pupper'].value_counts())
print(tweet_archive['puppo'].value_counts())

None     2259
doggo      97
Name: doggo, dtype: int64
None       2346
floofer      10
Name: floofer, dtype: int64
None      2099
pupper     257
Name: pupper, dtype: int64
None     2326
puppo      30
Name: puppo, dtype: int64


In [64]:
# Twitter Images
tweet_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [105]:
print(tweet_images.tweet_id.duplicated().sum())
print(tweet_images.tweet_id.isnull().sum())

0
0


In [107]:
# Tweets Dataframe
print(df_tweets.info())
print(df_tweets.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2331 non-null   int64
 1   retweet_count   2331 non-null   int64
 2   favorite_count  2331 non-null   int64
dtypes: int64(3)
memory usage: 54.8 KB
None
           tweet_id  retweet_count  favorite_count
count  2.331000e+03    2331.000000     2331.000000
mean   7.419079e+17    2636.776920     7410.796225
std    6.823170e+16    4459.108835    11507.204304
min    6.660209e+17       1.000000        0.000000
25%    6.782670e+17     535.000000     1288.000000
50%    7.182469e+17    1233.000000     3222.000000
75%    7.986692e+17    3061.500000     9066.500000
max    8.924206e+17   75764.000000   152970.000000


In [116]:
print(df_tweets.tweet_id.isnull().sum())
print(df_tweets.tweet_id.duplicated().sum())

0
0


In [128]:
print(df_tweets.retweet_count.isnull().sum())
print(df_tweets.retweet_count.sort_values())

0
279         1
328         2
1271        2
262         2
109         3
        ...  
65      39690
397     42668
517     55217
1055    56379
1015    75764
Name: retweet_count, Length: 2331, dtype: int64


In [129]:
print(df_tweets.favorite_count.isnull().sum())
print(df_tweets.favorite_count.sort_values())

0
913          0
415          0
409          0
406          0
404          0
         ...  
1055    113073
129     114272
517     118081
397     130464
1015    152970
Name: favorite_count, Length: 2331, dtype: int64


## Assessment Summary (from both visual and programmatic assessment)
### Quality
#### Tweets Archive
- includes replies, and retweets as seen by `in_reply_to_user_id` and `retweeted_status_user_id` columns (only want original tweets)
- +0000 is unnecessary in time column
- invalid data in `rating_denominator` column (anything other than 10)
- invalid data in `rating_numerator` column (anything less than 10 or greated than 20)
- `tweet id` column datatype datatype could be str instead of int (it is a name)
- `timestamp` column datatype should be datetime not str
- missing values in `name` column marked under 'None' and 'a' instead of NaN

#### Twitter Image
- `p1`, `p2`, and `p3` columns have underscores instead of spaces
- `tweet id` column datatype datatype could be str instead of int
- there are 2075 rows in this dataframe and 2356 rows in the archive dataframe

#### Tweets Dataframe
- there are 2331 rows in this dataframe and 2356 rows in the archive dataframe
- `tweet_id` column datatype could be str instead of int
- some retweet counts and favorite counts seem low
---
### Tidiness
#### Tweets Archive
- columns relating to retweet and reply data is irrelevant (we only want original tweets)
- `dog_stage` data is broken up into 4 columns when should be in 1

#### Twitter Image
- this dataframe is tidy

#### Tweets Dataframe
- this dataframe is tidy